content/posts/2022/kerberos-all-steps-workflow.md

```diff
@@ -0,0 +1,81 @@
+---
+title: "Easily enable Kerberos for all the steps of a workflow"
+date: 2022-11-30T14:30:00+01:00
+tags:
+  - kerberos
+---
+
+Up until now, it was only possible to enable Kerberos support for a workflow on
+a step-by-step basis. This changes with REANA 0.9.0, which makes it easy to
+enable Kerberos authentication for an entire workflow by using only a single
+hint in its specification file.
+
+<!--more-->
+
+As an example, let's consider a workflow that uses Kerberos to access a folder
+on an external restricted storage backend like EOS to publish results there.
+This is how the `reana.yaml` specification file might look like:
+
+```yaml {hl_lines=[10]}
+workflow:
+  type: serial
+  specification:
+    steps:
+      - name: myfirststep
+        #...
+      - name: mysecondstep
+        #...
+      - name: publish
+        kerberos: true
+        environment: "reanahub/krb5"
+        commands:
+          - mkdir -p /eos/home/j/johndoe/myanalysis-outputs
+          - cp myplots/*.png /eos/home/j/johndoe/myanalysis-outputs/
+```
+
+Setting `kerberos: true` on each step that needs it works well if just a few
+steps are involved, but it is not so good when dealing with a lot of them.
+Moreover, there are cases in which Kerberos authentication is needed even for
+parsing and validating the workflow specification, for example when using the
+Snakemake workflow engine with data objects living in a restricted data storage.
+
+For these reasons, it is now possible to enable Kerberos support for the whole
+workflow, by setting `kerberos: true` in the
+[`workflow.resources`](https://docs.reana.io/reference/reana-yaml/#reanayaml-workflow)
+section of the workflow's specification file:
+
+```yaml {hl_lines=["3-4"]}
+workflow:
+  type: serial
+  resources:
+    kerberos: true
+  specification:
+    steps:
+      - name: myfirststep
+        #...
+      - name: mysecondstep
+        #...
+      - name: publish
+        environment: "reanahub/krb5"
+        commands:
+          - mkdir -p /eos/home/j/johndoe/myanalysis-outputs
+          - cp myplots/*.png /eos/home/j/johndoe/myanalysis-outputs/
+```
+
+As you can see, setting up Kerberos in the `publish` step is not needed anymore.
```

Comment:
This technique is especially useful for Snakemake workflows that need to access files in restricted storage:

```
FIXME 
```

(Let's show the real example here. Otherwise people may think that we have simply implemented a hint propagation from top-level workflow section down to bottom-level job sections.  Therefore, showing a realistic Snakemake demo will help to clarify that this is indeed useful for the whole orchestration --  and as a by product it shall gently advertise that we have a nice Snakemake engine option in RIEANA.)

Reply:
@mdonadoni I can put a commit on top of your PR with the desired changes and fix all the formatting _etc_. OK with you?

Reply:
Good for me!